enhancements/oc/mirroring-manifests.md

```diff
@@ -0,0 +1,260 @@
+Users who intend to audit the gathered manifests should extract the gathered types from the manifests themselves.
+
+## Proposal
+
+### Applying directly to the target cluster
+
+As of 4.2 and 4.3:
+
+```console
+$ oc adm release mirror \
+>   --from=quay.io/openshift-release-dev/ocp-release:4.3.0-x86_64 \
+>   --to=registry.example.com/your/repository \
+>   --to-release-image=registry.example.com/your/repository:your-tag
+```
+
+will mirror the 4.3.0 release image and other images referenced by the release image into the mirror repository at registry.example.com.
+
+This enhancement adds an `--apply-manifests` option.
+When set, `oc` would also apply associated manifests to the cluster that `oc` connects to, with the manifest application beginning after the image mirror had completed.
+This enhancement would also allow `oc apply`'s `--overwrite` option to allow users to manage conflicts with resources that already existed in the target cluster.
+
+If `oc` encountered an error applying a manifest it should warn the user, and then either:
+
+* Immediately exit or
+* Continue to apply remaining manifest to the cluster.
+
+In either case, it would exit non-zero.
+
+### Pushing to disk
+
+As of 4.3 (with [this change][relative-to-dir-value] to support relative `--to-dir` values):
+
+```console
+$ oc adm release mirror --to-dir=mirror quay.io/openshift-release-dev/ocp-release:4.3.0-x86_64
+...
+Success
+Update image:  openshift/release:4.3.0
+
+To upload local images to a registry, run:
+
+    oc image mirror --from-dir=mirror file://openshift/release:4.3.0* REGISTRY/REPOSITORY
+```
+
+creates an output filesystem like:
+
+```console
+$ tree mirror
+mirror
+└── v2
+    └── openshift
+        └── release
+            ├── blobs
+            │   ├── sha256:005337dc1f6870934cf5efaf443786149e6a71041a91566ce1e16e78bce511a9
+            ...
+            │   └── sha256:ffc132fefac522ec38b1b22ecd67a5173565efb3e8929e47ef7e0b4ee7920adf
+            └── manifests
+                ├── 4.3.0 -> sha256:3a516480dfd68e0f87f702b4d7bdd6f6a0acfdac5cd2e9767b838ceede34d70d
+                ...
+                ├── 4.3.0-thanos -> sha256:42ab3f59d3e769e88af89c30304b9cc77a76b66960747e2e1a6f8db097420858
+                ├── sha256:00ec58112784d340179e045bc70399ac1bd509ae43c4411e934ced23d016b2a1
+                ...
+                └── sha256:ffa87c992c5a2c51290e6d67cb609f29893dd147c31cff639ff6785ae7a1cfe2
+
+5 directories, 514 files
+```
+
+This filesystem can then be applied to the local registry with the given `oc image mirror` command.
+
+With this enhancement, the output format would gain an additional directory `manifests` under the base directory, a sibling to the current `v2`, containing manifest files that can be applied to the local cluster with:
+
+```console
+$ oc apply -Rf mirror/manifests
+```
+
+The `manifests` directory may or may not contain additional subdirectories, so the recursive `-R` should be given for future-proofing.
```

Comment:
I'm a bit worried about future proofing here.  If we generated a kustomize output, the directories are implicit and apply would not be appropriate.  Let's actually spec this opposite way - for the current duration we should not use `-R` in order to preserve future extensibility.

Reply:
> If we generated a kustomize output, the directories are implicit and apply would not be appropriate...

Can you not apply kustomize with `oc apply`?  From [here][1], looks like it's:

```console
$ kustomize build ~/someApp | kubectl apply -f -
```

Is there not a way to recursively apply kustomize templates?  Maybe just:

```console
$ find mirror/kustomize -name '*.yaml' | while read X; do kustomize build "${X}" | oc apply -f -; done
```

I think that calls for a completely separate directory, which we don't need to pack into `kubernetes-manifests`, not dropping the recursive apply.

[1]: https://github.com/kubernetes-sigs/kustomize#1-make-a-kustomization-file

Reply:
I really don't want recursion unless we have a reason.  Don't be broad in what you accept if you can be narrow.  If we decide to do recursion, we should have it.

Reply:
> Don't be broad in what you accept if you can be narrow...

I think I addressed the kustomize example in my earlier comment.  Can you provide an example where a user would have a Kube config manifest generated by this mirror command which they would not want to push into the target cluster?  I can't think of an example, and recursive application requires less admin thought (although admins that like thinking are free to inspect manifests and push non-recursively if they want).  If we land this with the recursive suggestion (and propagate that suggestion into openshift-docs), and then later come up with a manifest that we think is not a no-brainer for pushing, we can create a new directory `optional-config` or whatever, so I'm not worried about painting ourselves into a corner.

Reply:
I am still not convinced that it is a big deal to have `-R`. I think we are trying to address a problem which can complicate the design without much gain in user experience as we do not know if users would need sub-directories or use them. The use case we are trying to address is a simple one where we do not need sub directories. Also I would assume those will be advance users who does not go with the default workflow and we do not need to care about advance uses at this point of time of this enhancement request.

Reply:
> Also I would assume those will be advance users who does not go with the default workflow and we do not need to care about advance uses at this point of time of this enhancement request.

And for these advanced users, we [already say][1]:

> Users who intend to audit the gathered configuration manifests should extract the gathered types from the configuration manifests themselves.

so the path for folks who chose not to apply recursively is already clear.

[1]: https://github.com/openshift/enhancements/commit/884166a05806719c559ea3f74a7c5370508e150e#diff-13e449411659f84a0f8827c8a909c727R51

Reply:
This is getting confusing. Can you please explain why you think the `config` directory may or may not contain additional sub-directories.  If you have an example (abstract example is fine) that would be good. 
```
The `config` directory may or may not contain additional subdirectories, so the recursive `-R` should be given for future-proofing.
```

Reply:
> Can you please explain why you think the `config` directory may or may not contain additional sub-directories.

Initially, we'll put signature ConfigMap manifest(s) in the `config` directory.  No subdirectories.  Say we keep growing the types of manifests and decide for some reason that it would help to collect a bunch of update-related manifests in a subdirectory `config/update`.  With the recursive-apply default, users don't have to care about shifts like that.  The "may or may not" is a reminder that any users designing flows around this UX should not make assumptions about whether or not subdirectories will exist.

Reply:
Took a stab at [a more explicit comparison between recursive and non-recursive][1].

[1]: https://github.com/openshift/enhancements/pull/188#issuecomment-606165632